event-store/src/adapters/amqp/emitter.rs

```diff
@@ -0,0 +1,182 @@
+//! AMQP emitter implementation
+
+use adapters::EmitterAdapter;
+use futures::future::{ok, Future};
+use futures::Stream;
+use lapin::channel::{
+    BasicConsumeOptions, BasicProperties, BasicPublishOptions, Channel, ExchangeDeclareOptions,
+    QueueBindOptions, QueueDeclareOptions,
+};
+use lapin::client::{Client, ConnectionOptions};
+use lapin::types::FieldTable;
+use serde_json;
+use std::io;
+use std::net::SocketAddr;
+use std::str;
+use std::sync::mpsc;
+use std::time::Duration;
+use tokio;
+use tokio::net::TcpStream;
+use tokio::runtime::Runtime;
+use Event;
+use EventData;
+use Events;
+
+/// AMQP emitter
+pub struct AMQPEmitterAdapter {
+    channel: Channel<TcpStream>,
+    exchange: String,
+    namespace: String,
+}
+
+impl AMQPEmitterAdapter {
+    /// Create a new AMQPEmiterAdapter
+    pub fn new(
+        uri: SocketAddr,
+        _exchange: String,
+        namespace: String,
```

Comment:
Can `exchange` and `namespace` become not `String`s now we have derive?

Reply:
Good call. `namespace` is no longer required as it's provided by the event derive.

The exchange is still required as it's a requirement specific of the AMQPAdapter.